### PR TITLE
Wire up sync for testing, with an initial set of tests for invoke()

### PR DESCRIPTION
**Pull request by david-martin**
start must be called to setup sync mongoclient, queues and workers.
start will always be called by fh-mbaas-api on initialisation.
start is not required to be called for testing

### CODE REVIEW

File: lib/sync/index.js [hunk 1]
```diff
@@ -132,6 +135,7 @@ function start(cb) {
     },
     function startSyncScheduler(callback) {
       syncScheduler = new SyncScheduler();
```
Comment (fh-qube) [inline lib/sync/index.js]: MAJOR Add the "var" keyword to this declaration of "syncScheduler". [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AS2703)

File: lib/sync/index.js [hunk 2]
```diff
@@ -17,26 +18,31 @@ function toJSON(dataset_id, returnData, cb) {
...
-    return callback("unknown_fn : " + fn, null);
+    return callback('invalid fn parameter provided in params "' + fn + '"', null);
   }
 
   return fnHandler(dataset_id, params, callback);
 }
 
+// Initialise cloud data sync service for specified dataset.
 function init(dataset_id, options, cb) {
   syncUtil.doLog(dataset_id, 'info', 'init');
```
Comment (fh-qube) [inline lib/sync/index.js]: MINOR Remove the useless trailing whitespaces at the end of this line. [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3ATrailingWhitespace)

File: lib/sync/index.js [hunk 3]
```diff
@@ -89,34 +99,27 @@ function setLogLevel(dataset_id, params, cb) {
...
   return cb(null, {});
 }
 
-/* Synchronise the individual records for a dataset */
-function syncRecords(dataset_id, params, callback) {
+// Synchronise the individual records for a dataset
+function syncRecords(dataset_id, params, cb) {
   syncUtil.doLog(dataset_id, 'info', 'syncRecords');
 
   // TODO
```
Comment (fh-qube) [inline lib/sync/index.js]: INFO Complete the task associated to this TODO comment. [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AS1135)

File: lib/sync/index.js [hunk 4]
```diff
@@ -53,6 +60,7 @@ function stop(dataset_id, cb) {
   cb(null, {});
 }
 
+// Stop cloud data sync service for ALL datasets.
 function stopAll(cb) {
   syncUtil.doLog(syncUtil.SYNC_LOGGER, 'info', 'stopAll');
```
Comment (fh-qube) [inline lib/sync/index.js]: MINOR Remove the useless trailing whitespaces at the end of this line. [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3ATrailingWhitespace)

File: lib/sync/index.js [hunk 5]
```diff
@@ -17,26 +18,31 @@ function toJSON(dataset_id, returnData, cb) {
   return cb(null, {});
 }
 
+// Invoke the Sync Server.
 function invoke(dataset_id, params, callback) {
   syncUtil.doLog(dataset_id, 'info', 'invoke');
 
+  if (arguments.length < 3) throw new Error('invoke requires 3 arguments');
```
Comment (fh-qube) [inline lib/sync/index.js]: MINOR Reformat the code to have only one statement per line. [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AOneStatementPerLine)

File: lib/sync/index.js [hunk 6]
```diff
@@ -45,6 +51,7 @@ function init(dataset_id, options, cb) {
   cb(null, {});
 }
 
+// Stop cloud data sync for the specified dataset_id.
 function stop(dataset_id, cb) {
   syncUtil.doLog(dataset_id, 'info', 'stop');
```
Comment (fh-qube) [inline lib/sync/index.js]: MINOR Remove the useless trailing whitespaces at the end of this line. [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3ATrailingWhitespace)

File: lib/sync/index.js [hunk 7]
```diff
@@ -89,34 +99,27 @@ function setLogLevel(dataset_id, params, cb) {
   }
 }
 
-function sync(dataset_id, params, callback) {
+function sync(dataset_id, params, cb) {
   syncUtil.doLog(dataset_id, 'info', 'sync');
 
   // TODO
```
Comment (fh-qube) [inline lib/sync/index.js]: INFO Complete the task associated to this TODO comment. [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AS1135)

**Comment by fh-qube** [comment]
SonarQube analysis reported 13 issues:
* MAJOR 1 major
* MINOR 4 minor
* INFO 8 info

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* INFO [Complete the task associated to this TODO comment.](https://github.com/feedhenry/fh-mbaas-api/blob/c42a221087d4804c53e2f4f2d271f81f3aa194a3/lib/sync/index.js#L75) [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AS1135)
* INFO [Complete the task associated to this TODO comment.](https://github.com/feedhenry/fh-mbaas-api/blob/c42a221087d4804c53e2f4f2d271f81f3aa194a3/lib/sync/index.js#L16) [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AS1135)
* INFO [Complete the task associated to this TODO comment.](https://github.com/feedhenry/fh-mbaas-api/blob/c42a221087d4804c53e2f4f2d271f81f3aa194a3/lib/sync/index.js#L58) [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AS1135)
* INFO [Complete the task associated to this TODO comment.](https://github.com/feedhenry/fh-mbaas-api/blob/c42a221087d4804c53e2f4f2d271f81f3aa194a3/lib/sync/index.js#L49) [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AS1135)
* INFO [Complete the task associated to this TODO comment.](https://github.com/feedhenry/fh-mbaas-api/blob/c42a221087d4804c53e2f4f2d271f81f3aa194a3/lib/sync/index.js#L67) [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AS1135)
* INFO [Complete the task associated to this TODO comment.](https://github.com/feedhenry/fh-mbaas-api/blob/c42a221087d4804c53e2f4f2d271f81f3aa194a3/lib/sync/index.js#L85) [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AS1135)

**Comment by david-martin** [comment]
@wei-lee Some inital wiring up of tests.
I'm thinking we'll need a combination of unit tests and tests that require redis & mongodb

File: Gruntfile.js [hunk 8]
```diff
@@ -16,23 +16,24 @@ module.exports = function(grunt) {
...
+    // './test/test_sync.js',
+    './test/sync/test_index.js'
+    // './test/test_sync_handlers.js',
+    // './test/test_sync_datasetmodel.js',
+    // './test/test_sync_utils.js',
+    // './test/test_fhauth.js',
+    // './test/test_init.js',
+    // './test/test_log.js',
+    // './test/test_fhpush.js'
   ];
```
Comment (wei-lee) [inline Gruntfile.js]: maybe we can create a separate task to just test the sync related tests?

File: lib/sync/index.js [hunk 9]
```diff
@@ -136,49 +158,45 @@ function start(cb) {
       return callback();
     },
     function startSyncScheduler(callback) {
-      syncScheduler = new SyncScheduler();
+      syncScheduler = new SyncScheduler(redisClient);
```
Comment (fh-qube) [inline lib/sync/index.js]: MAJOR Add the "var" keyword to this declaration of "syncScheduler". [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AS2703)

File: lib/sync/index.js [hunk 10]
```diff
@@ -89,38 +104,45 @@ function setLogLevel(dataset_id, params, cb) {
...
+  async.parallel([
+    function connectToMongoDB(callback) {
+      MongoClient.connect(mongoDBConnectionUrl, callback);
+    },
+    function connectToRedis(callback) {
+      var client = redis.createClient(redisUrl);
+      return callback(null, client);
+    }
+  ], function(err, results) {
+    if (err) return cb(err);
```
Comment (fh-qube) [inline lib/sync/index.js]: MINOR Reformat the code to have only one statement per line. [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AOneStatementPerLine)

**Comment by fh-qube** [comment]
SonarQube analysis reported 14 issues:
* MAJOR 1 major
* MINOR 5 minor
* INFO 8 info

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* INFO [Complete the task associated to this TODO comment.](https://github.com/feedhenry/fh-mbaas-api/blob/20f306a51423db29af0ad74d0f31cc95781d2e09/lib/sync/index.js#L18) [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AS1135)
* INFO [Complete the task associated to this TODO comment.](https://github.com/feedhenry/fh-mbaas-api/blob/20f306a51423db29af0ad74d0f31cc95781d2e09/lib/sync/index.js#L90) [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AS1135)
* INFO [Complete the task associated to this TODO comment.](https://github.com/feedhenry/fh-mbaas-api/blob/20f306a51423db29af0ad74d0f31cc95781d2e09/lib/sync/index.js#L72) [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AS1135)
* INFO [Complete the task associated to this TODO comment.](https://github.com/feedhenry/fh-mbaas-api/blob/20f306a51423db29af0ad74d0f31cc95781d2e09/lib/sync/index.js#L80) [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AS1135)
* INFO [Complete the task associated to this TODO comment.](https://github.com/feedhenry/fh-mbaas-api/blob/20f306a51423db29af0ad74d0f31cc95781d2e09/lib/sync/index.js#L63) [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AS1135)
* INFO [Complete the task associated to this TODO comment.](https://github.com/feedhenry/fh-mbaas-api/blob/20f306a51423db29af0ad74d0f31cc95781d2e09/lib/sync/index.js#L54) [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AS1135)

**Pull request merged**